File: src/flycns/rng.py
"""Counter-based randomness: the same (seed, neuron, step) gives the same number in Python, TypeScript and WGSL.

The generator is MurmurHash3_x86_32 (Austin Appleby's public-domain hash) of the 8-byte little-endian key
(neuron, step), with the run's seed as the hash seed. Every operation is a 32-bit unsigned multiply, add, xor, shift
or rotation, which JavaScript (``Math.imul``, ``>>> 0``), WGSL (``u32`` arithmetic wraps) and NumPy (``uint32`` wraps)
compute identically, and any MurmurHash3 implementation is an oracle for it. A Poisson event of rate ``r`` at step
size ``dt`` occurs when the hash is below ``floor(r * dt * 2**32)``, so an event stream needs no generator state, can
be evaluated for any neuron at any step in any order, and reproduces bit for bit across implementations.
"""

from __future__ import annotations

import numpy as np

_C1 = np.uint32(0xCC9E2D51)
_C2 = np.uint32(0x1B873593)
_ROUND = np.uint32(0xE6546B64)
_FIVE = np.uint32(5)
_MIX_A = np.uint32(0x85EBCA6B)
_MIX_B = np.uint32(0xC2B2AE35)
_KEY_BYTES = np.uint32(8)


def _rotl(x: np.ndarray, r: int) -> np.ndarray:
    return (x << np.uint32(r)) | (x >> np.uint32(32 - r))


def fmix32(h: np.ndarray) -> np.ndarray:
    """MurmurHash3's 32-bit finaliser, on uint32 arrays."""
    h = np.asarray(h, dtype=np.uint32)
    with np.errstate(over="ignore"):
        h = h ^ (h >> np.uint32(16))
        h = h * _MIX_A
        h = h ^ (h >> np.uint32(13))
        h = h * _MIX_B
        h = h ^ (h >> np.uint32(16))
    return h


def _block(h: np.ndarray, k: np.ndarray) -> np.ndarray:
    """One 4-byte block of MurmurHash3_x86_32's body."""
    k = _rotl(k * _C1, 15) * _C2
    return _rotl(h ^ k, 13) * _FIVE + _ROUND


def hash3(seed: int, neuron: np.ndarray, step: int | np.ndarray) -> np.ndarray:
    """MurmurHash3_x86_32 of the little-endian bytes of (neuron, step), seeded with ``seed``; a uint32 array."""
    neuron = np.asarray(neuron, dtype=np.uint32)
    step = np.asarray(step, dtype=np.uint32)
    with np.errstate(over="ignore"):
        h = np.full(np.broadcast_shapes(neuron.shape, step.shape), seed & 0xFFFFFFFF, dtype=np.uint32)
        h = _block(h, neuron)
        h = _block(h, step)
        return fmix32(h ^ _KEY_BYTES)


def event_threshold(rate_hz: float, dt_s: float) -> int:
    """The uint32 threshold below which a hash counts as an event of probability ``rate * dt``."""
    p = min(max(rate_hz * dt_s, 0.0), 1.0)
    return min(int(p * 2**32), 2**32 - 1)


def poisson_events(seed: int, neurons: np.ndarray, step: int, threshold: int | np.ndarray) -> np.ndarray:
    """Boolean mask: which of ``neurons`` receive an event at ``step``."""
    return hash3(seed, neurons, step) < np.asarray(threshold, dtype=np.uint64)

File: ts/src/rng.ts
/**
 * Counter-based randomness, the same numbers as `flycns.rng`: MurmurHash3_x86_32 of the 8-byte little-endian key
 * (neuron, step), seeded with the run's seed. Every operation is a 32-bit unsigned multiply (`Math.imul`), add, xor,
 * shift or rotation, so the result equals NumPy's uint32 arithmetic and the WGSL kernel's u32 arithmetic bit for bit.
 */

const C1 = 0xcc9e2d51;
const C2 = 0x1b873593;
const ROUND = 0xe6546b64;
const MIX_A = 0x85ebca6b;
const MIX_B = 0xc2b2ae35;
const KEY_BYTES = 8;

function rotl(x: number, r: number): number {
  return ((x << r) | (x >>> (32 - r))) >>> 0;
}

function block(h: number, k: number): number {
  k = Math.imul(k, C1) >>> 0;
  k = rotl(k, 15);
  k = Math.imul(k, C2) >>> 0;
  h = (h ^ k) >>> 0;
  h = rotl(h, 13);
  return (Math.imul(h, 5) + ROUND) >>> 0;
}

/** MurmurHash3's 32-bit finaliser. */
export function fmix32(h: number): number {
  h = (h ^ (h >>> 16)) >>> 0;
  h = Math.imul(h, MIX_A) >>> 0;
  h = (h ^ (h >>> 13)) >>> 0;
  h = Math.imul(h, MIX_B) >>> 0;
  return (h ^ (h >>> 16)) >>> 0;
}

/** MurmurHash3_x86_32 of (neuron, step) seeded with `seed`: an unsigned 32-bit integer. */
export function hash3(seed: number, neuron: number, step: number): number {
  let h = seed >>> 0;
  h = block(h, neuron >>> 0);
  h = block(h, step >>> 0);
  return fmix32((h ^ KEY_BYTES) >>> 0);
}

/**
 * The threshold below which a hash counts as an event of probability `rate * dt`; `rateHz * dtS` is computed here
 * as the reference's `event_threshold` computes it (the caller passes `dtS` already divided).
 */
export function eventThreshold(rateHz: number, dtS: number): number {
  const p = Math.min(Math.max(rateHz * dtS, 0), 1);
  return Math.min(Math.floor(p * 4294967296), 4294967295);
}

/** Whether `neuron` receives an event at `step` under `threshold`. */
export function poissonEvent(seed: number, neuron: number, step: number, threshold: number): boolean {
  return hash3(seed, neuron, step) < threshold;
}
